google-cloud-spanner/src/test/java/com/google/cloud/spanner/DefaultBenchmark.java

```diff
@@ -0,0 +1,246 @@
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ThreadLocalRandom;
+import java.util.concurrent.TimeUnit;
+import org.openjdk.jmh.annotations.AuxCounters;
+import org.openjdk.jmh.annotations.Benchmark;
+import org.openjdk.jmh.annotations.BenchmarkMode;
+import org.openjdk.jmh.annotations.Fork;
+import org.openjdk.jmh.annotations.Level;
+import org.openjdk.jmh.annotations.Measurement;
+import org.openjdk.jmh.annotations.Mode;
+import org.openjdk.jmh.annotations.OutputTimeUnit;
+import org.openjdk.jmh.annotations.Scope;
+import org.openjdk.jmh.annotations.Setup;
+import org.openjdk.jmh.annotations.State;
+import org.openjdk.jmh.annotations.TearDown;
+import org.openjdk.jmh.annotations.Warmup;
+
+/**
+ * Benchmarks for measuring existing latencies of various APIs using the Java Client. The benchmarks
+ * are bound to the Maven profile `benchmark` and can be executed like this: <code>
+ * mvn clean test -DskipTests -Pbenchmark -Dbenchmark.name=DefaultBenchmark
+ * </code> Test Table Schema :
+ *
+ * <p>CREATE TABLE FOO ( id INT64 NOT NULL, BAZ INT64, BAR INT64, ) PRIMARY KEY(id);
+ *
+ * <p>Below are a few considerations here: 1. We use all default options for this test because that
+ * is what most customers would be using. 2. The test schema uses a numeric primary key. To ensure
+ * that the reads/updates are distributed across a large query space, we insert 10^5 records.
+ * Utility at {@link BenchmarkingDataLoaderUtility} can be used for loading data. 3. For queries, we
+ * make sure that the query is sampled randomly across a large query space. This ensure we don't
+ * cause hot-spots. 4. For avoid cold start issues, we execute 1 query/update and ignore its latency
+ * from the final reported metrics.
+ */
+@BenchmarkMode(Mode.AverageTime)
+@Fork(value = 1, warmups = 0)
+@Measurement(batchSize = 1, iterations = 1, timeUnit = TimeUnit.MILLISECONDS)
+@OutputTimeUnit(TimeUnit.SECONDS)
+@Warmup(iterations = 1)
+public class DefaultBenchmark {
+  private static final String SELECT_QUERY = "SELECT ID FROM FOO WHERE ID = @id";
+  private static final String UPDATE_QUERY = "UPDATE FOO SET BAR=1 WHERE ID = @id";
+  private static final String ID_COLUMN_NAME = "id";
+  private static final int PARALLEL_THREADS = 1;
+  private static final int TOTAL_READS_PER_THREAD = 12000;
+  private static final int TOTAL_WRITES_PER_THREAD = 4000;
+  private static final int WARMUP_TRANSACTIONS = 1;
+  private static final int RANDOM_SEARCH_SPACE = 99999;
+
+  @State(Scope.Thread)
+  @AuxCounters(org.openjdk.jmh.annotations.AuxCounters.Type.EVENTS)
+  public static class BenchmarkState {
+
+    private static final String INSTANCE_ID = "";
+    private static final String DATABASE_ID = "";
+    private static final String SERVER_URL = "";
+    private Spanner spanner;
+    private DatabaseClientImpl client;
+
+    @Setup(Level.Invocation)
```

Comment:
Are we sure that this is the right level for these benchmarks? The Javadoc for this level is quite telling (bring your own sword!): https://javadoc.io/static/org.openjdk.jmh/jmh-core/1.1.1/org/openjdk/jmh/annotations/Level.html#Invocation

Reply:
My requirement here was to re-initialise the session pool for every invocation. But I think we can change it to `@Iteration` given the cost of setting up the session would anyway get excluded in the top 1% results.